### PR TITLE
Dynamic cert support, rewrite installers

### LINKED ISSUE #142: Ubuntu 17.04 (Zesty Zapus) - Unescaped left brace in regex is deprecated, passed through in regex; marked by
I have this problem on install.
Sorry if duplicated issue.

Thanks.

### LINKED ISSUE #491: No HTTPS support macOS Catalina
### Update

A stop-gap patch has been applied to 2.0.11.  Please download it here: https://qz.io/download

----

To apply a patch to previous versions, follow these instructions:
https://youtu.be/jghL1PDukME

https://support.apple.com/en-us/HT210176

> Learn about new security requirements for TLS server certificates in iOS 13 and macOS 10.15.
>
> All TLS server certificates must comply with these new security requirements in iOS 13 and macOS 10.15:
>
> * TLS server certificates and issuing CAs using RSA keys must use key sizes greater than or equal to 2048 bits. Certificates using RSA key sizes smaller than 2048 bits are no longer trusted for TLS.
> * TLS server certificates and issuing CAs must use a hash algorithm from the SHA-2 family in the signature algorithm. SHA-1 signed certificates are no longer trusted for TLS.
> * TLS server certificates must present the DNS name of the server in the Subject Alternative Name extension of the certificate. DNS names in the CommonName of a certificate are no longer trusted.
> * Additionally, all TLS server certificates issued after July 1, 2019 (as indicated in the NotBefore field of the certificate) must follow these guidelines:
> 
> * TLS server certificates must contain an ExtendedKeyUsage (EKU) extension containing the id-kp-serverAuth OID.
> * ### TLS server certificates must have a validity period of `825 days` or fewer (as expressed in the `NotBefore` and `NotAfter` fields of the certificate).
>
> Connections to TLS servers violating these new requirements will fail and may cause network > failures, apps to fail, and websites to not load in Safari in iOS 13 and macOS 10.15.

### LINKED ISSUE #494: Mask tray icons wrong color Windows 10
Windows 10 appears to have a "Custom" option to change the way Dark mode behaves:

image

With the correct combination, the Tray Icon is white on white (or black on black) making it hard to see/find.

### LINKED ISSUE #496: Fix autostart logic
I believe when branches were synced, 2.1 accidentally got 2.0's cleanup tasks. 

Due to order of operations macOS seems to be the only platform that suffers, it deletes the entry right after it creates it. 🤦‍♂ 

https://github.com/qzind/tray/blob/7eebf5f3f3ba53a0f8b9353881538baff56ed2ba/ant/apple/apple-postinstall.sh.in#L27-L30

https://github.com/qzind/tray/blob/7eebf5f3f3ba53a0f8b9353881538baff56ed2ba/ant/linux/linux-installer.sh.in#L112-L113

https://github.com/qzind/tray/blob/7eebf5f3f3ba53a0f8b9353881538baff56ed2ba/ant/windows/windows-packager.nsi.in#L57-L58

### PR DESCRIPTION
**Pull request by tresf**
Moves all installation tasks to Java, allows certs to renew automatically.  There's a bunch of enhancements in this PR, but I'll try to high-level them... 

Download: https://github.com/tresf/tray/releases/tag/v2.10-RC8

### features
* 825-day SSL certs, fixing Catalina support (comply with [Ballot 193](https://cabforum.org/2017/03/17/ballot-193-825-day-certificate-lifetimes/), closes #491)
* Support soft-reload of properties and keystore (closes #275 (supercede), related #447)
* Multiple Firefox version support (closes #142)
* Experimental support for LetsEncrypt cert renewals (untested, related #73)
* Consistent Java search behavior (related #468)
* Prefer AdoptOpenJDK to Oracle
* Also closes #494 (needs backport), closes #496

### certgen
Out with the old `widows-keygen.js`, `linux-keygen.sh` and `apple-keygen.sh`, in with the new, standardized scripts.
```bash
java -jar qz-tray.jar certgen --key "privkey.pem" --cert "fullchain.pem"
java -jar qz-tray.jar certgen --pfx "mycert.pfx" --pass "12345"
java -jar qz-tray.jar certgen --host "192.168.1.5"
java -jar qz-tray.jar certgen --hosts "domain1;domain2;..."
```
... called as sudo or as a regular user. It'll install the cert in the user's certstore now, meaning ssl from IntelliJ is now possible.

---

### uninstall
```bash
java -jar qz-tray.jar uninstall
```
... called as sudo/admin of course

---

### advanced
Internally we now use java for all install tasks... 
```bash
java -jar qz-tray.jar preinstall # kills all instances of qz
java -jar qz-tray.jar install    # deploys qz to a sane location
```
... called as sudo/admin of course

---

### launchers
Platform launchers have been improved to pass parameters directly to qz-tray.jar...
```bash
C:\Program Files\qz-tray.exe "certgen"
/opt/qz-tray/qz-tray "certgen"
/Applications/QZ\ Tray.app/QZ\ Tray "certgen"
```

### CODE REVIEW

File: src/qz/utils/JsonWriter.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            if (baseVal == null) {
+                //add new key
+                base.put(key, mergeVal);
+            } else if (baseVal instanceof JSONObject && mergeVal instanceof JSONObject) {
+                //deep copy sub-keys
+                merge((JSONObject)baseVal, (JSONObject)mergeVal, overwrite);
+            } else if (overwrite) {
+                //force new key val if existing and allowed
+                base.put(key, mergeVal);
+            } else if (baseVal instanceof JSONArray && mergeVal instanceof JSONArray) {
```
Comment (bberenz) [inline src/qz/utils/JsonWriter.java]: I believe you mentioned a repeat entry bug in arrays for this I still need to fix. 
Adding a comment so I remember to look into it.

File: src/qz/utils/SystemUtilities.java [hunk 2]
```diff
@@ -168,6 +171,8 @@ public static boolean isWindows() {
         return (OS_NAME.contains("win"));
     }
 
+    public static boolean isWindowsXP() { return OS_NAME.contains("win") && OS_NAME.contains("xp"); }
```
Comment (bberenz) [inline src/qz/utils/SystemUtilities.java]: Why not under WindowsUtilities class?

File: src/qz/utils/CommandParser.java [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+import qz.installer.TaskKiller;
+import qz.installer.certificate.PropertiesLoader;
+
+import java.io.File;
+import java.util.*;
+
+import static qz.common.Constants.*;
+import static qz.utils.CommandParser.ExitStatus.*;
+
+public class CommandParser {
```
Comment (bberenz) [inline src/qz/utils/CommandParser.java]: Something like ArgumentParser would probably be a better name for what this is doing.

File: src/qz/utils/CommandParser.java [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Gets the argument value immediately following a command
+     * @throws NullCommandException
+     */
+    public String argValue(String ... matches) throws NullCommandException {
```
Comment (bberenz) [inline src/qz/utils/CommandParser.java]: This class is just for args, so we could just call this "value" or "valueOf"

File: src/qz/utils/CommandParser.java [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    }
+
+    public int getExitCode() {
+        return exitStatus.getCode();
+    }
+
+    /**
+     * Gets the requested flag status
+     */
+    public boolean argFlag(String ... matches) {
```
Comment (bberenz) [inline src/qz/utils/CommandParser.java]: Similar for this, could call this "hasFlag" or "isSet", would be clearer to its purpose too

File: ant/linux/linux.properties [hunk 6]
```diff
@@ -1,41 +0,0 @@
-# Linux build properties
```
[the remark this answers is not in the document]
Comment (tresf) [reply ant/linux/linux.properties]: Removed via c5f038b.

File: src/qz/common/AboutInfo.java [hunk 7]
```diff
@@ -67,13 +65,13 @@ private static JSONObject product() throws JSONException {
...
 
-    private static JSONObject socket(KeyStore keystore, String domain) throws JSONException {
+    private static JSONObject socket(PropertiesLoader propertiesLoader, String domain) throws JSONException {
         JSONObject socket = new JSONObject();
 
         socket
                 .put("domain", domain)
                 .put("secureProtocol", "wss")
-                .put("securePort", keystore == null? "none":PrintSocketServer.getSecurePortInUse())
+                .put("securePort", propertiesLoader.isSslActive() ? "none":PrintSocketServer.getSecurePortInUse())
```
[the remark this answers is not in the document]
Comment (tresf) [reply src/qz/common/AboutInfo.java]: Good catch.  Fixed via https://github.com/qzind/tray/pull/504/commits/c5f038bed6487a249aa840459729def7d00d35fa#diff-232539db024c6b187aa18d6e5c274831R74.

File: src/qz/common/TrayManager.java [hunk 8]
```diff
@@ -221,40 +211,68 @@ private void addMenuItems() {
...
+        zipLogs.setMnemonic(KeyEvent.VK_Z);
+        zipLogs.addActionListener(e -> FileUtilities.zipLogs());
+        diagnosticMenu.add(zipLogs);
 
         JMenuItem desktopItem = new JMenuItem("Create Desktop shortcut", iconCache.getIcon(IconCache.Icon.DESKTOP_ICON));
         desktopItem.setMnemonic(KeyEvent.VK_D);
         desktopItem.addActionListener(desktopListener());
 
+        anonymousItem = new JCheckBoxMenuItem("Block anonymous requests");
+        anonymousItem.setToolTipText("Blocks all requests that do no contain a valid certificate/signature");
```
[the remark this answers is not in the document]
Comment (tresf) [reply src/qz/common/TrayManager.java]: Fixed via https://github.com/qzind/tray/pull/504/commits/c5f038bed6487a249aa840459729def7d00d35fa#diff-a953e83b62c6bf3017ac9df374ffb42dR263.

File: src/qz/installer/Installer.java [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+        TaskKiller.killAll();
+        getInstance();
+        log.info("Uninstalling from {}", instance.getDestination());
+        instance.removeSharedDirectory()
+                .removeSystemSettings()
+                .removeCerts();
+    }
+
+    private static Installer instance;
+    public static Installer getInstance() {
```
[the remark this answers is not in the document]
Comment (tresf) [reply src/qz/installer/Installer.java]: Done via https://github.com/qzind/tray/pull/504/commits/c5f038bed6487a249aa840459729def7d00d35fa#diff-74389ccba4a9cf4b0150413dbd68bb04R80.

File: src/qz/exception/NullCommandException.java [hunk 10]
```diff
@@ -1,6 +1,9 @@
 package qz.exception;
 
 public class NullCommandException extends javax.print.PrintException {
```
[the remark this answers is not in the document]
Comment (tresf) [reply src/qz/exception/NullCommandException.java]: Added `MissingArgException` via https://github.com/qzind/tray/pull/504/commits/c5f038bed6487a249aa840459729def7d00d35fa#diff-46670d13c3e6b804cb676b4c488c9514R1.

File: src/qz/installer/certificate/WindowsCertificateInstaller.java [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+
+    public void setInstallType(Installer.PrivilegeLevel type) {
+        this.type = type;
+    }
+
+    public Installer.PrivilegeLevel getInstallType() {
+        return type;
+    }
+
+    public static String parseNextLine(BufferedReader reader) throws IOException {
```
[the remark this answers is not in the document]
Comment (tresf) [reply src/qz/installer/certificate/WindowsCertificateInstaller.java]: Done via https://github.com/qzind/tray/pull/504/commits/c5f038bed6487a249aa840459729def7d00d35fa#diff-92cd93fd523d2a96d91b395b7313b30eR125.

File: src/qz/installer/certificate/firefox/FirefoxCertificateInstaller.java [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+
+    public static boolean honorsPolicy(AppLocator app) {
+        if (app.getVersion() == null) {
+            log.warn("Firefox-compatible browser was found {}, but no version information is available", app.getPath());
+            return false;
+        }
+        if(SystemUtilities.isWindows()) {
+            return app.getVersion().greaterThanOrEqualTo(MAC_POLICY_VERSION);
+        } else if (SystemUtilities.isMac()) {
+            return app.getVersion().greaterThanOrEqualTo(WINDOWS_POLICY_VERSION);
```
[the remark this answers is not in the document]
Comment (tresf) [reply src/qz/installer/certificate/firefox/FirefoxCertificateInstaller.java]: Fixed.

File: src/qz/installer/certificate/PropertiesLoader.java [hunk 13]
```diff
@@ -0,0 +1,415 @@
...
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.util.*;
+
+import static qz.utils.FileUtilities.*;
+import static qz.installer.certificate.KeyPairWrapper.Type.*;
+
+/**
+ * Stores and maintains reading and writing of certificate related files
+ */
+public class PropertiesLoader {
```
[the remark this answers is not in the document]
Comment (tresf) [reply src/qz/installer/certificate/PropertiesLoader.java]: Changed via c5f038b.

File: src/qz/utils/CommandParser.java (hunk 3 above)
Comment (tresf) [reply src/qz/utils/CommandParser.java]: Done, renamed to `ArgParser`.

File: src/qz/installer/certificate/KeyPairWrapper.java [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+
+    public X509Certificate getCert() {
+        return cert;
+    }
+
+    public PrivateKey getKey() {
+        return key;
+    }
+
+    public String getPass() {
```
[the remark this answers is not in the document]
Comment (tresf) [reply src/qz/installer/certificate/KeyPairWrapper.java]: Changed to `getPasswordString` per recommendation.

File: src/qz/utils/CommandParser.java (hunk 4 above)
Comment (tresf) [reply src/qz/utils/CommandParser.java]: Changed to `valueOf` per recommendation.

File: src/qz/utils/CommandParser.java (hunk 5 above)
Comment (tresf) [reply src/qz/utils/CommandParser.java]: changed to `hasFlag` per recommendation.

**Comment by tresf** [comment]
All comments addressed.   @bberenz three comments still remain unresolved.

File: src/qz/utils/JsonWriter.java (hunk 1 above)
Comment (bberenz) [reply src/qz/utils/JsonWriter.java]: Pushed a fix
In regards to the formatting anomaly you mentioned before, it looks like a bug in the library when only 1 key is present in an object, they have a special n==1 case that forgets to add newlines.. Everything formats correctly for any objects with 2 or more keys.

File: src/qz/installer/MacInstaller.java [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+        return destination;
+    }
+
+    public Installer addUserSettings() { return this; }
+
+    public Installer addSystemSettings() { return this; }
+    public Installer removeSystemSettings() {
+        // Remove startup entry
+        File dest = new File(String.format("/Library/LaunchAgents/%s.plist", PACKAGE_NAME));
+        dest.delete();
```
[the remark this answers is not in the document]
Comment (bberenz) [reply src/qz/installer/MacInstaller.java]: >  Some functions use Apache's FileUtils.deleteDirectory() which actually throws an IOException if not found

Ah, I see, that must have been what I was seeing elsewhere.
If we don't care if this fails then it should be fine as is.

File: src/qz/installer/certificate/ExpiryTask.java [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+        // Check for expiration
+        ExpiryState state = getExpiry(loader.getSslKeyPair().getCert());
+        switch(state) {
+            case EXPIRING:
+            case EXPIRED:
+                log.info("Certificate ExpiryState {}, renewing/reloading...", state);
+                switch(certProvider) {
+                    case INTERNAL:
+                        if(renewInternalCert()) {
+                            getExpiry();
```
[the remark this answers is not in the document]
Comment (bberenz) [reply src/qz/installer/certificate/ExpiryTask.java]: That's fine, I just wasn't sure if we were only calling it for the logging or if there was something missing in its use.

File: src/qz/utils/JsonWriter.java (hunk 1 above)
Comment (tresf) [reply src/qz/utils/JsonWriter.java]: Did you test your theory?  I added another element, and this is what I got... It's still not formatting properly.

```json
{
  "policies": {"Certificates": {"ImportEnterpriseRoots": true}},
  "indentFix": -99999
}
```

Then I wrapped it in an array and it's still wrong... 

```json
[
  {"policies": {"Certificates": {"ImportEnterpriseRoots": true}}},
  {"indentFix": -99999}
]
```

Comment (tresf) [reply src/qz/utils/JsonWriter.java]: Nevermind, editing the class proves the original point.  Working on a patch now...

Comment (tresf) [reply src/qz/utils/JsonWriter.java]: Updated Jettison with a custom patch.  Indentation looks good now.  I just removed the `length() == 1` block completely. 

Sent request to both upstream projects (neither appear to be maintained)
https://github.com/codehaus/jettison/pull/10
https://github.com/jettison-json/jettison/pull/31

**Comment by tresf** [comment]
All conversations have been resolved.  This is ready for merge.

**Comment by lite1979** [comment]
Raw>Raster Print> HTML printing fails with AdoptOpenJDK 11 on Windows 10

First attempt resulted in an error complaining Java FX did not start, but I did not get the error after that.Error message coincided with logs finally giving up, but they logged "Waiting for JavaFX to start" for almost 90 seconds before letting the processor go; might be a problem if it happened on a machine with less RAM, but I was able to send other print jobs while it was waiting.

**Comment by tresf** [comment]
@lite1979 thanks.  Patched via https://github.com/qzind/tray/issues/520 directly on the 2.1 branch.  This seems to be a long-standing 2.1 bug since we've added Java 11 support.

**Comment by tresf** [comment]
Testing this branch with a LetsEncrypt certificate and it works, but the following error occurs:

```
[ERROR] 2019-11-20 12:55:35,084 @ qz.installer.Installer:?
	Something went wrong obtaining the certificate.  HTTPS will fail.
org.bouncycastle.util.io.pem.PemGenerationException: encoding exception: unknown object passed - can't encode.
	at org.bouncycastle.openssl.MiscPEMGenerator.generate(Unknown Source)
	at qz.installer.certificate.CertificateManager.writeCert(Unknown Source)
	at qz.installer.Installer.certGen(Unknown Source)
	at qz.ws.PrintSocketServer.main(Unknown Source)
Caused by: org.bouncycastle.util.io.pem.PemGenerationException: unknown object passed - can't encode.
	at org.bouncycastle.openssl.MiscPEMGenerator.createPemObject(Unknown Source)
	... 4 more
```

**Comment by tresf** [comment]
Print-Server steps updated with 2.1.0-RC9+ commands. https://github.com/qzind/tray/wiki/Print-Server

**Comment by tresf** [comment]
There seems to be an edge-case issue where Firefox may not be detected on MacOS during the installation process.

The problem seems to stem from the [Spotlight service not listing Firefox as installed](https://apple.stackexchange.com/questions/248660/apps-dont-show-up-on-spotlight-search). You can test this by calling the following command:

```bash
system_profiler SPApplicationsDataType -xml |grep "Firefox"
```

The fix is to stop and start the Spotlight service using the following command:
```bash
sudo mdutil -a -i off
sudo mdutil -a -i on
```

Then test for Firefox again:

```bash
system_profiler SPApplicationsDataType -xml |grep "Firefox"
```

Should output:

```xml
        			<string>Firefox</string>
				<string>Firefox 72.0.2</string>
				<string>/Applications/Firefox.app</string>
```

**Pull request merged**